shell/platform/android/io/flutter/view/AccessibilityBridge.java

```diff
@@ -35,35 +63,143 @@
+    // Flutter supports custom accessibility actions via {@code customSemanticsActions} within
+    // a {@code Semantics} widget, https://docs.flutter.io/flutter/widgets/Semantics-class.html.
+    // {@code customAccessibilityActions} are an Android-side cache of all custom accessibility
+    // types declared within the running Flutter app.
+    //
+    // Custom accessibility actions are comprised of only a few fields, and therefore it is likely
+    // that a given app may define the same custom accessibility action many times. Identical
+    // custom accessibility actions are de-duped such that {@code customAccessibilityActions} only
+    // caches unique custom accessibility actions.
+    //
+    // See the Android documentation for custom accessibility actions:
+    // https://developer.android.com/reference/android/view/accessibility/AccessibilityNodeInfo.AccessibilityAction
+    //
+    // See the Flutter documentation for the Semantics widget:
+    // https://docs.flutter.io/flutter/widgets/Semantics-class.html
+    @NonNull
+    private final Map<Integer, CustomAccessibilityAction> customAccessibilityActions;
+
+    // TODO(mattcarroll): what is the source of truth for this supposed to be? Is it
+    //                    whether or not the OS has accessibility enabled? Or do other
+    //                    factors influence this value?
     private boolean accessibilityEnabled = false;
-    private SemanticsObject a11yFocusedObject;
-    private SemanticsObject inputFocusedObject;
-    private SemanticsObject hoveredObject;
+
+    // The {@code SemanticsNode} within Flutter that currently has the focus of Android's
+    // accessibility system.
+    @Nullable
+    private SemanticsNode accessibilityFocusedSemanticsNode;
+
+    // The {@code SemanticsNode} within Flutter that currently has the focus of Android's input
+    // system.
+    //
+    // Input focus is independent of accessibility focus. It is possible that accessibility focus
+    // and input focus target the same {@code SemanticsNode}, but it is also possible that one
+    // {@code SemanticsNode} has input focus while a different {@code SemanticsNode} has
+    // accessibility focus. For example, a user may use a D-Pad to navigate to a text field, giving
+    // it accessibility focus, and then enable input on that text field, giving it input focus. Then
+    // the user moves the accessibility focus to a nearby label to get info about the label, while
+    // maintaining input focus on the original text field.
+    @Nullable
+    private SemanticsNode inputFocusedSemanticsNode;
+
+    // The widget within Flutter that currently sits beneath a cursor, e.g,
+    // beneath a stylus or mouse cursor.
+    @Nullable
+    private SemanticsNode hoveredObject;
+
+    // TODO(mattcarroll): what is this? the term "route" sounds navigation oriented...
```

Comment:
So Routing,

on Android, whenever the user switches activities an "announcement" is made of the new activity name and accessibility focus is moved to the first focusable accessibility node. This is done by emitting a AccessibilityEvent.TYPE_WINDOW_STATE_CHANGED event with a id which corresponds to the first node.

This idea is heavily tied to the idea of routes, and page transitions. I believe the a11y name for this would be edge-driven semantics. The fact that it is activity driven is an implementation detail on Android, the announcements should happen on anything which drastically changes the screen, such as a new "route", dialog, or modal.

To do this accurately, we keep a stack of ids which correspond to the router stack on the flutter side. When a route is pushed/pop'd we use the track'd ids to determine which semantics node id to send and which label to use when reporting.

Reply:
Updated the navigation stack comment.  Still need to trace the purpose of previous route.